pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,633 @@
+
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+	if lease != nil {
+		if lease.Annotations[ElectedByAnnotationName] == controllerName {
+			klog.Infof("reconcile for lease namespace=%q, name=%q", lease.Namespace, lease.Name)
+		} else if isLeaseExpired(lease) {
+			clone := lease.DeepCopy()
+			if lease.Annotations[ElectedByAnnotationName] == controllerName && lease.Spec.HolderIdentity != nil && clone.Spec.RenewTime != nil && clone.Spec.LeaseDurationSeconds != nil && clone.Spec.AcquireTime != nil {
+				delete(clone.Annotations, EndOfTermAnnotationName)
+				clone.ObjectMeta.Annotations[ElectedByAnnotationName] = controllerName
+				clone.Spec.HolderIdentity = nil
+				clone.Spec.RenewTime = nil
+				clone.Spec.LeaseDurationSeconds = nil
+				clone.Spec.AcquireTime = nil
+				_, err := c.leaseClient.Leases(clone.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+	}
+
+	err = c.scheduleElection(leaderLeaseID{namespace: namespace, name: name})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Controller) scheduleElection(leaderLeaseID leaderLeaseID) error {
+	// TODO: Create additional options for skipping elections for leases that opt out of CLE
+	if leaderLeaseID.name == "leader-election-controller" {
+		return nil
+	}
+	klog.Infof("scheduleElection")
+	// TODO: add a set to track ongoing elections and avoid requesting an election if one has already been kicked off?
+	// This might not be needed..
+	c.electionCh <- election{
+		leaderLeaseID: leaderLeaseID,
+		electionStart: time.Now(),
+	}
+	return nil
+}
+
+func (c *Controller) runElection(ctx context.Context, leaderLeaseID leaderLeaseID) error {
+	klog.Infof("runElection %q %q", leaderLeaseID.namespace, leaderLeaseID.name)
+	candidates, err := c.listCandidates(leaderLeaseID)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Optimize to short circult here if candidates and leader are unchanged
+	electee0 := pickBestLeader(candidates)
+	if electee0 == nil {
+		return nil
+	}
+
+	ct := c.initializeCT(leaderLeaseID, len(candidates))
+
+	for _, candidate := range candidates {
+		clone := candidate.DeepCopy()
+		if clone.Annotations == nil {
+			clone.Annotations = make(map[string]string)
+		}
+		clone.Annotations[AckAnnotationName] = "Pending"
+		_, err := c.leaseCandidateClient.LeaseCandidates(clone.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	select {
+	case <-ct.done:
+	case <-time.After(5 * time.Second):
+		klog.Infof("Warning: election timed out after 5s for lease %s/%s", leaderLeaseID.namespace, leaderLeaseID.name)
+	}
+
+	var properCandidates []*v1alpha1.LeaseCandidate
+	for _, candidate := range candidates {
+		if ct.candidates[candidate.Name] {
+			properCandidates = append(properCandidates, candidate)
+		}
+	}
+
+	electee := pickBestLeader(properCandidates)
+	if electee == nil {
+		return nil
```

Comment:
klog not leader found? maybe at V(2) or greater if is something that can be noisy